modin/pandas/groupby.py

```diff
@@ -171,10 +173,76 @@ def idxmax(self):
     def ndim(self):
         return 2  # ndim is always 2 for DataFrames
 
-    def shift(self, periods=1, freq=None, axis=0):
-        return self._default_to_pandas(
-            lambda df: df.shift(periods=periods, freq=freq, axis=axis)
-        )
+    def shift(self, periods=1, freq=None, axis=0, fill_value=None):
+        """
+        Shift each group by periods observations.
+
+        If freq is passed, the index will be increased using the periods and the freq.
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Number of periods to shift.
+        freq : str, optional
+            Frequency string.
+        axis : axis to shift, default 0
+            Shift direction.
+        fill_value : optional
+            The scalar value to use for newly introduced missing values.
+
+        Returns
+        -------
+        Series or DataFrame
+            Object shifted within each group.
+        """
+
+        def _shift(periods, freq, axis, fill_value, is_set_nan_rows=True):
+            from .dataframe import DataFrame
+
+            result = self._df.shift(periods, freq, axis, fill_value)
+
+            if is_set_nan_rows:
+                if (
+                    isinstance(self._by, BaseQueryCompiler)
+                    and set(self._by.columns).issubset(list(self._df.columns))
+                    and self._by.isna().to_numpy().any()
```

Comment:
why do we need to convert it to `numpy`? why `self._by.isna().any()` doesn't work?

Reply:
We do that because `self._by.isna().any()` returns `Series` or `DataFrame` and `to_numpy().any()` returns boolean value

Reply:
converting to numpy seems expensive, can we do some combinations of `squeeze` and `any` to get bool value or do `by.isnull().sum().sum() != 0`, can you please do a time measurement of these approaches and pick best?

Reply:
I've tested 4 implementations. Looks like `to_numpy` approach is the best way.

nrows=1kk, ncols=100
```
to_numpy 1.011391719803214
using sum 1.1405104491859674
using sumsum 1.265167985111475
using values 3.0229647886008024
```
nrows=100k, ncols=100
```
to_numpy 0.07831187918782234
using sum 0.10889025591313839
using sumsum 0.12529431097209454
using values 0.2087058462202549
```
nrows=10k, ncols=100
```
to_numpy 0.015122836455702782
using sum 0.03983925096690655
using sumsum 0.03679628483951092
using values 0.02024906873703003
```
<details><summary>script</summary>

```python
import modin.pandas as pd
import numpy as np
from timeit import default_timer as timer

NROWS = 10_000
NCOLS = 100
RAND_LOW = 0
RAND_HIGH = 10

random_state = np.random.RandomState(seed=42)
data = {"col{}".format(int(i)): random_state.randint(
            RAND_LOW, RAND_HIGH, size=(NROWS)
        )
        for i in range(NCOLS)
    }

md_df = pd.DataFrame(data)
md_df.loc[8,'col3'] = np.nan
md_df.loc[15,'col9'] = np.nan
md_df.loc[1000,'col80'] = np.nan


t1 = timer()
res = md_df.isna().to_numpy().any()
print("to_numpy", timer() - t1)

t1 = timer()
res = md_df.isna().any().sum()
print("using sum", timer() - t1)

t1 = timer()
res = md_df.isna().sum().sum()
print("using sumsum", timer() - t1)

t1 = timer()
res = md_df.values.any()
print("using values", timer() - t1)

```
<details>

Reply:
I think we can avoid a call `to_numpy` in such way: `any(self._by.isna())`. @prutskov , can you check does it make sense?

Reply:
@YarShev  We can't use your approach, because `type(self._by.isna())` is `type(query_compiler)` (not iterable object).

Reply:
I am not sure what is faster but we could try this: `any(col.any() for _, col in DataFrame(query_compiler=self._by.isna()).items())`.

Reply:
@YarShev iterating on modin DataFrame are usually slow, but it would be good if we measure that approach too.


> We do that because self._by.isna().any() returns Series or DataFrame and to_numpy().any() returns boolean value

`self._by.isna().any().any().to_pandas().squeeze()` will return an boolean that we need

Reply:
Results are presented below. `to_numpy` is faster.
nrows=10k, ncols=100
```
to_numpy 0.016001993790268898
using sum 0.0426933728158474
using sumsum 0.041401373222470284
using any() 0.05819194205105305
```
nrows=100k, ncols=100
```
to_numpy 0.07790362276136875
using sum 0.10964982956647873
using sumsum 0.12710685655474663
using any() 0.19313380308449268
```
nrows=1kk, ncols=100
```
to_numpy 1.0193398352712393
using sum 1.1495637483894825
using sumsum 1.3674616161733866
using any() 2.1003225687891245
```

<details><summary>script</summary>

```python
import modin.pandas as pd
import numpy as np
from timeit import default_timer as timer

NROWS = 1000_000
NCOLS = 100
RAND_LOW = 0
RAND_HIGH = 10

random_state = np.random.RandomState(seed=42)
data = {"col{}".format(int(i)): random_state.randint(
            RAND_LOW, RAND_HIGH, size=(NROWS)
        )
        for i in range(NCOLS)
    }

md_df = pd.DataFrame(data)
md_df.loc[8,'col3'] = np.nan
md_df.loc[15,'col9'] = np.nan
md_df.loc[1000,'col80'] = np.nan

md_qc = md_df._query_compiler

t1 = timer()
res = md_qc.isna().to_numpy().any()
print("to_numpy", timer() - t1)

t1 = timer()
res = md_qc.isna().any().sum()
print("using sum", timer() - t1)

t1 = timer()
res = md_qc.isna().sum().sum()
print("using sumsum", timer() - t1)

t1 = timer()
res = any(col.any() for _, col in pd.DataFrame(query_compiler=md_qc.isna()).items())
print("using any()", timer() - t1)

```
</details>

Reply:
@prutskov Im getting this results for `.any().any().to_pandas().squeeze()` approach:

```
# shape (5_000_000, 100)
to_numpy 1.7891445886343718
using any 0.4493228103965521
```

<details><summary> Script </summary>


```python
import modin.pandas as pd
import numpy as np
from timeit import default_timer as timer

NROWS = 5_000_000
NCOLS = 100
RAND_LOW = 0
RAND_HIGH = 10

random_state = np.random.RandomState(seed=42)
data = {"col{}".format(int(i)): random_state.randint(
            RAND_LOW, RAND_HIGH, size=(NROWS)
        )
        for i in range(NCOLS)
    }

md_df = pd.DataFrame(data)
md_df.loc[8,'col3'] = np.nan
md_df.loc[15,'col9'] = np.nan
md_df.loc[1000,'col80'] = np.nan

md_qc = md_df._query_compiler

t1 = timer()
res = md_qc.isna().to_numpy().any()
print("to_numpy", timer() - t1)

t1 = timer()
res = md_qc.isna().any().any().to_pandas().squeeze()
print("using any", timer() - t1)

```


</details>

Reply:
@dchigarev some fix `self._by.isna().any().any(axis=1).to_pandas().squeeze()`. It's slower than `to_numpy`.

nrows=1kk, ncols=100
```
using squeeze() 1.1528270151466131
```
nrows=100k, ncols=100
```
using squeeze() 0.10938875377178192
```
nrows=10k, ncols=100
```
using squeeze() 0.03129498101770878
```

Reply:
@dchigarev @YarShev I fixed problems with my environment, correct results are presented below. It's strange because time of `using sum` is decreased with increasing of data size.
nrows=10k, ncols=100
```
to_numpy 0.1964775286614895
using sum 1.806968130171299
using sumsum 0.4314875267446041
using any() 1.1656176634132862
using squeeze() 1.0665200389921665
```
nrows=100k, ncols=100
```
to_numpy 0.21689333207905293
using sum 1.797909015789628
using sumsum 0.40592130459845066
using any() 1.0854842234402895
using squeeze() 1.030497558414936
```
nrows=1kk, ncols=100
```
to_numpy 1.4334983844310045
using sum 0.6739870477467775
using sumsum 0.48233154974877834
using any() 1.2144871186465025
using squeeze() 1.095459209755063
```
nrows=10kk, ncols=100
```
to_numpy 2.3792185336351395
using sum 0.5216180328279734
using sumsum 0.9957287088036537
using any() 1.1807466875761747
using squeeze() 1.1313506197184324
```